### PR TITLE
ENH: pd.MultiIndex.get_loc(np.nan)

### LINKED ISSUE #19132: pd.MultiIndex.get_loc(np.nan)
#### Code Sample, a copy-pastable example if possible

```python
In [2]: mi = pd.MultiIndex.from_product([[np.nan, 1]] * 3)

In [3]: mi.get_loc(np.nan)
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/home/pietro/nobackup/repo/pandas/pandas/core/indexes/base.py in get_loc(self, key, method, tolerance)
   2513             try:
-> 2514                 return self._engine.get_loc(key)
   2515             except KeyError:

/home/pietro/nobackup/repo/pandas/pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc (pandas/_libs/index.c:5834)()

/home/pietro/nobackup/repo/pandas/pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc (pandas/_libs/index.c:5639)()

/home/pietro/nobackup/repo/pandas/pandas/_libs/index_class_helper.pxi in pandas._libs.index.Int64Engine._check_type (pandas/_libs/index.c:19008)()

KeyError: nan

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
<ipython-input-3-641c024055e9> in <module>()
----> 1 mi.get_loc(np.nan)

/home/pietro/nobackup/repo/pandas/pandas/core/indexes/multi.py in get_loc(self, key, method)
   2121 
   2122         if not isinstance(key, tuple):
-> 2123             loc = self._get_level_indexer(key, level=0)
   2124             return _maybe_to_slice(loc)
   2125 

/home/pietro/nobackup/repo/pandas/pandas/core/indexes/multi.py in _get_level_indexer(self, key, level, indexer)
   2411         else:
   2412 
-> 2413             loc = level_index.get_loc(key)
   2414             if isinstance(loc, slice):
   2415                 return loc

/home/pietro/nobackup/repo/pandas/pandas/core/indexes/base.py in get_loc(self, key, method, tolerance)
   2514                 return self._engine.get_loc(key)
   2515             except KeyError:
-> 2516                 return self._engine.get_loc(self._maybe_cast_indexer(key))
   2517 
   2518         indexer = self.get_indexer([key], method=method, tolerance=tolerance)

/home/pietro/nobackup/repo/pandas/pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc (pandas/_libs/index.c:5834)()

/home/pietro/nobackup/repo/pandas/pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc (pandas/_libs/index.c:5639)()

/home/pietro/nobackup/repo/pandas/pandas/_libs/index_class_helper.pxi in pandas._libs.index.Int64Engine._check_type (pandas/_libs/index.c:19008)()

KeyError: nan

```

#### Problem description

Opening a separate issue from #18485 (full label containing NaN) because, differently from that, this is _not_ fixed by #19074.

xref #5286 because they are probably worth testing together.

#### Expected Output

``slice(0, 4, None)``

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-5-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: en_GB.UTF-8

pandas: 0.22.0.dev0+448.gdbec3c92e
pytest: 3.0.6
pip: 9.0.1
setuptools: 33.1.1
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.18.1
pyarrow: None
xarray: None
IPython: 5.2.2
sphinx: None
patsy: 0.4.1+dev
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: 1.2.0
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: None
lxml: 3.7.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.8
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by proost**
MultiIndex.get_loc could not find nan with values including missing
values as a input.

Background: In `MultiIndex`, missing value is denoted by -1 in codes and doesn't exist in `self.levels`

So, could not find NA value in `self.levels`.

Before PR xref #28783

- [x] closes #19132
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -2601,7 +2601,7 @@ def _partial_tup_index(self, tup, side="left"):
                     loc -= 1
                 return start + section.searchsorted(loc, side=side)
 
-            idx = lev.get_loc(lab)
+            idx = lev.get_loc(lab) if not isna(lab) else -1
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: getting on a level should itself return -1, so why is this needed?

File: pandas/core/indexes/multi.py [hunk 2]
```diff
@@ -2707,7 +2707,8 @@ def _maybe_to_slice(loc):
         loc = np.arange(start, stop, dtype="int64")
 
         for i, k in enumerate(follow_key, len(lead_key)):
-            mask = self.codes[i][loc] == self.levels[i].get_loc(k)
+            code = self.levels[i].get_loc(k) if not isna(k) else -1
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: same

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/multi.py (hunk 1 above)
Comment (proost) [reply pandas/core/indexes/multi.py]: @jreback  Are you sure? Because  where code will be changed is in indexes.base or _libs.index. So effect of change quite a big.

File: pandas/tests/indexes/multi/test_indexing.py [hunk 3]
```diff
@@ -437,3 +437,30 @@ def test_timestamp_multiindex_indexer():
...
+def test_get_indexer_with_missing_value():
+    # issue 19132
+    idx = MultiIndex.from_arrays([[1, np.nan, 2], [3, 4, 5]])
+    result = idx.get_indexer([1, np.nan, 2])
+    expected = np.array([-1, -1, -1], dtype="int32")
+    tm.assert_numpy_array_equal(result.astype("int32"), expected)
+
+    result = idx.get_indexer([(np.nan, 4)])
+    expected = np.array([1], dtype="int32")
+    tm.assert_numpy_array_equal(result.astype("int32"), expected)
```
Comment (jbrockmendel) [inline pandas/tests/indexes/multi/test_indexing.py]: why use int32?  if you used int64 for expected you wouldnt need to cast here, right?

File: pandas/tests/indexes/multi/test_indexing.py [hunk 4]
```diff
@@ -437,3 +437,30 @@ def test_timestamp_multiindex_indexer():
...
+    idx = MultiIndex.from_arrays([[np.nan, 1, 2, np.nan]])
+    expected = np.array([True, False, False, True])
+    tm.assert_numpy_array_equal(idx.get_loc(np.nan), expected)
+
+    idx = MultiIndex.from_product([[np.nan, 1]] * 3)
+    expected = slice(2, 4, None)
+    assert idx.get_loc((np.nan, 1)) == expected
+
+
+def test_get_indexer_with_missing_value():
```
Comment (jbrockmendel) [inline pandas/tests/indexes/multi/test_indexing.py]: testing get_loc and get_indexer are good, but we dont expect users to see those directly.  is there any user-facing behavior that this changes that should be tested?

File: pandas/core/indexes/multi.py [hunk 5]
```diff
@@ -2628,13 +2628,25 @@ def _partial_tup_index(self, tup, side="left"):
...
 
-            idx = lev.get_loc(lab)
+            idx = self._get_loc_single_index_wrapper(lev, lab)
             if k < n - 1:
                 end = start + section.searchsorted(idx, side="right")
                 start = start + section.searchsorted(idx, side="left")
             else:
                 return start + section.searchsorted(idx, side=side)
 
+    def _get_loc_single_index_wrapper(self, level_index, key):
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: name these parameters how they are called (level, label), pls add type annotations and a full-docstring.

File: pandas/core/indexes/multi.py [hunk 6]
```diff
@@ -2628,13 +2628,25 @@ def _partial_tup_index(self, tup, side="left"):
...
+        """
+        single index get_loc wrapper function for multi index.
+        Depending on index type, Handling NA value is different using get_loc.
+        But in multi index, NA values is denoted as missing by -1.
+        If key is NA value, location of index unify as -1.
+        """
+        if not is_list_like(key) and isna(key):
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: how is this *every* list-like? isn't it supposed to be a scalar?

File: pandas/core/indexes/multi.py [hunk 7]
```diff
@@ -2628,13 +2628,25 @@ def _partial_tup_index(self, tup, side="left"):
...
+            return -1
+        else:
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: rather than having changing MultiIndex like this, can you just update .get_loc on Index? e.g. it correctly returns -1 for a null input?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/multi.py (hunk 7 above)
Comment (proost) [reply pandas/core/indexes/multi.py]: @jreback 
Well, That is more than this issue. For example,
```
In [2]: idx = pd.Index(['a','b',None,'c','d']) 
   ...: idx.get_loc(None)                                                       
Out[2]: 2
```
Like this case, some Index `.get_loc` can find NA value. I thought it meant NA value doesn't consider as missing. But if return -1 for a null input, then it regards as missing. From then, this PR get off topic. this PR deals with Multiindex's missing and single index missing both. I think for single index separate issue.

File: pandas/core/indexes/multi.py (hunk 6 above)
Comment (proost) [reply pandas/core/indexes/multi.py]: @jreback 
I change into check scalar or not. In `PeriodIndex`, all key except when key is tuple, call single index get_loc. For a NA value type check, it is inevitable.

File: pandas/core/indexes/multi.py [hunk 8]
```diff
@@ -2618,13 +2618,45 @@ def _partial_tup_index(self, tup, side="left"):
...
+            If key is NA value, loc is -1
+            Else, location of key in index
+
+        See Also
+        --------
+        Index.get_loc : The get_loc method for (single-level) index.
+
+        Notes
+        -----
+        Depending on Index type, Handling NA value is different using get_loc.
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: this note is confusing and irrelevant, pls remove

File: pandas/core/indexes/multi.py [hunk 9]
```diff
@@ -2618,13 +2618,45 @@ def _partial_tup_index(self, tup, side="left"):
...
+            idx = self._get_loc_single_level_index_wrapper(lev, lab)
             if k < n - 1:
                 end = start + section.searchsorted(idx, side="right")
                 start = start + section.searchsorted(idx, side="left")
             else:
                 return start + section.searchsorted(idx, side=side)
 
+    def _get_loc_single_level_index_wrapper(self, level_index: Index, key) -> int:
+        """
+        Wrapper function for MultiIndex.
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: this first line is not necessary

File: pandas/core/indexes/multi.py [hunk 10]
```diff
@@ -2618,13 +2618,45 @@ def _partial_tup_index(self, tup, side="left"):
...
+        If key is NA value, location of index unify as -1.
+
+        Parameters
+        ----------
+        level_index: Index
+                    Single level index in MultiIndex
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: this 2nd line is unecessary

File: pandas/core/indexes/multi.py [hunk 11]
```diff
@@ -2618,13 +2618,45 @@ def _partial_tup_index(self, tup, side="left"):
...
 
-            idx = lev.get_loc(lab)
+            idx = self._get_loc_single_level_index_wrapper(lev, lab)
             if k < n - 1:
                 end = start + section.searchsorted(idx, side="right")
                 start = start + section.searchsorted(idx, side="left")
             else:
                 return start + section.searchsorted(idx, side=side)
 
+    def _get_loc_single_level_index_wrapper(self, level_index: Index, key) -> int:
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: I would drop the _wrapper from the name

File: pandas/tests/indexes/multi/test_indexing.py (hunk 4 above)
Comment (jreback) [reply pandas/tests/indexes/multi/test_indexing.py]: can you respond to this, is there any user facing change aside from .get_loc itself, IOW does user facing indexing change?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/multi/test_indexing.py (hunk 4 above)
Comment (proost) [reply pandas/tests/indexes/multi/test_indexing.py]: @jreback 
`.get_slice_bound`, `__contains__`,`slice_indexer`, `slice_locs` affect from this change.
I add test case too.

**Comment by proost** [comment]
@jbrockmendel @jreback 
`get_slice_bound`, `__contains__`, `slice_indexer`, `slice_locs` changes from this change. In short, these functions can index NA value. I think, this is better than before. Because As we treat NA value like a non-NA value, we can index more accurately when NA value is in 'MultiIndex'

If,
```
idx = MultiIndex.from_arrays([[np.nan, "a", "b"], ["c", "d", "e"]])
```

case1 : "get_slice_bound" 
```
idx.get_slice_bound(np.nan, side="left", kind="ix")
Before:   raise exception
After:  0
```

case2 : "slice_indexer"
```
idx.slice_indexer(start=np.nan, end="a")   
Before: raise exception
After: slice(0, 2, None)
```

case3: "slice_locs"
```
idx.slice_locs(start=np.nan, end=("a","d"))  
Before: raise exception
After: (0,2)
```

case4: "__contains__"
```
assert np.nan in idx
Before: raise exception
After: Not raise exception
```
problem is when this PR is merged,   if "__contains__" is true, but "isin" can be false. because "isin" not affect from this change. so if this PR is merged, need to change "isin" for a consistency.

**Comment by jreback** [comment]
> problem is when this PR is merged, if __contains__ is true, but "isin" can be false. because "isin" not affect from this change. so if this PR is merged, need to change "isin" for a consistency.

can you show what you mean here?

**Comment by proost** [comment]
@jreback 

```
idx = MultiIndex.from_arrays([[np.nan, "a", "b"], ["c", "d", "e"]])
assert np.nan in idx
```
not raise exception. But
```
idx.isin([np.nan],level=0)
array([False, False, False])
```
`__contains__` says NA value in level 0 index, But `isin` says doesn't. 

So  If this PR is merged, change `isin` for consistency might be good.

**Comment by jreback** [comment]
can you merge master and will look again

**Comment by jreback** [comment]
> So If this PR is merged, change isin for consistency might be good.

ok this PR looks good. can you add in support for .isin() in this PR? I think it would make sense to merge these simultaneously (e.g. you can also build on this one if that is easier)

**Comment by proost** [comment]
@jreback 
Okay, I change `.isin` and add tests

File: pandas/tests/indexes/multi/test_indexing.py [hunk 12]
```diff
@@ -437,3 +437,91 @@ def test_timestamp_multiindex_indexer():
...
+            [np.nan, 4, 5],
+            np.array([-1, -1, -1], dtype="int64"),
+        ),
+    ],
+)
+def test_get_indexer_with_missing_value(index_arr, labels, expected):
+    # issue 19132
+    idx = MultiIndex.from_arrays(index_arr)
+    result = idx.get_indexer(labels)
+    tm.assert_numpy_array_equal(result.astype("int64"), expected)
```
Comment (jreback) [inline pandas/tests/indexes/multi/test_indexing.py]: rather than making everything int64, can you leave off the int dtypes (or use intp).

also structure the tests like

result = 
expected = 
assert_......(result, expected)

easier to read and test

File: pandas/core/indexes/multi.py [hunk 13]
```diff
@@ -2575,13 +2575,38 @@ def _partial_tup_index(self, tup, side="left"):
...
 
-            idx = lev.get_loc(lab)
+            idx = self._get_loc_single_level_index(lev, lab)
             if k < n - 1:
                 end = start + section.searchsorted(idx, side="right")
                 start = start + section.searchsorted(idx, side="left")
             else:
                 return start + section.searchsorted(idx, side=side)
 
+    def _get_loc_single_level_index(self, level_index: Index, key) -> int:
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: can you try to type key here

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. some comments, ping on green.

@toobaz @jbrockmendel @TomAugspurger if you'd care to have a look

File: pandas/core/indexes/multi.py [hunk 14]
```diff
@@ -1,5 +1,6 @@
 import datetime
 from sys import getsizeof
+from typing import Hashable, List, Optional
 from typing import Hashable, List, Optional, Sequence, Union
```
Comment (jbrockmendel) [inline pandas/core/indexes/multi.py]: this line looks repeated

File: pandas/core/indexes/multi.py [hunk 15]
```diff
@@ -3377,14 +3405,11 @@ def isin(self, values, level=None):
...
-            level_codes = self.codes[num]
+            levs = self.get_level_values(num)
 
-            sought_labels = levs.isin(values).nonzero()[0]
             if levs.size == 0:
-                return np.zeros(len(level_codes), dtype=np.bool_)
-            else:
-                return np.lib.arraysetops.in1d(level_codes, sought_labels)
+                return np.zeros(len(levs), dtype=np.bool_)
+            return levs.isin(values)
```
Comment (jbrockmendel) [inline pandas/core/indexes/multi.py]: the edits from L3408 down to here look like they are just nice cleanups independent of the rest of this PR.  is that accurate?

File: pandas/tests/indexes/multi/test_contains.py [hunk 16]
```diff
@@ -98,3 +98,27 @@ def test_isin_level_kwarg():
...
+    assert np.nan not in idx
+    assert (1, np.nan) in idx
+
+
+@pytest.mark.parametrize(
+    "labels,expected,level",
+    [
+        ([("b", np.nan)], np.array([False, False, True]), None,),
+        ([np.nan, "a"], np.array([True, True, False]), 0),
+        (["d", np.nan], np.array([False, True, True]), 1),
```
Comment (jbrockmendel) [inline pandas/tests/indexes/multi/test_contains.py]: is the issue specific to np.nan, or are there other NA values worth testing?

Comment (proost) [reply pandas/tests/indexes/multi/test_contains.py]: @jbrockmendel 
xref #30677. Yes. it is specific to np.nan. In "Index", np.nan , np.NaT, None are discernible not denoted by NA value. So if MultiIndex mixs with np.nan, np.NaT, None all together, result of ".isin" are different from what we know.

File: pandas/core/indexes/multi.py (hunk 15 above)
Comment (proost) [reply pandas/core/indexes/multi.py]: @jbrockmendel 
For NA values, someone fixes #30677, then more accurate. "Index.isin" has a bug nonetheless in terms of checking NA value can be possible, This is more accurate

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @proost very nice!
